ccl/multiregionccl: TestMrSystemDatabase failed

ccl/multiregionccl.TestMrSystemDatabase [failed](https://tanzanite.cluster.engflow.com/invocations/default/be232a8b-c406-45e5-a00a-fdbe1a5bf38f?testReportRun=3&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpcmVnaW9uY2NsOm11bHRpcmVnaW9uY2NsX3Rlc3Q=) on master @ [347cdc76d4c5abb2e872f325e944337a46b5883f](https://github.com/cockroachdb/cockroach/commits/347cdc76d4c5abb2e872f325e944337a46b5883f):


```
=== RUN   TestMrSystemDatabase
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestMrSystemDatabase2776683919
    test_log_scope.go:81: use -show-logs to present logs inline
    multiregion_system_table_test.go:381: -- test log scope end --
test logs left over in: outputs.zip/logTestMrSystemDatabase2776683919
--- FAIL: TestMrSystemDatabase (111.23s)
=== RUN   TestMrSystemDatabase/Sqlliveness_system_database
    multiregion_system_table_test.go:114: error scanning '&{0xc0195b0a20 <nil>}': pq: unsupported comparison: bytes to crdb_internal_region
    --- FAIL: TestMrSystemDatabase/Sqlliveness_system_database (0.19s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMrSystemDatabase.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-38068

ccl/multiregionccl.TestMrSystemDatabase [failed](https://tanzanite.cluster.engflow.com/invocations/default/8f9acc1e-0be6-4041-a068-6149e3253e6b?testReportRun=2&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpcmVnaW9uY2NsOm11bHRpcmVnaW9uY2NsX3Rlc3Q=) on master @ [2f498d0abcf4130e5c872b73200086d6a3af9f02](https://github.com/cockroachdb/cockroach/commits/2f498d0abcf4130e5c872b73200086d6a3af9f02):


```
=== RUN   TestMrSystemDatabase
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestMrSystemDatabase3383108453
    test_log_scope.go:81: use -show-logs to present logs inline
    multiregion_system_table_test.go:381: -- test log scope end --
test logs left over in: outputs.zip/logTestMrSystemDatabase3383108453
--- FAIL: TestMrSystemDatabase (274.12s)
=== RUN   TestMrSystemDatabase/Sqlliveness_system_database
    multiregion_system_table_test.go:114: error scanning '&{0xc02757d9e0 <nil>}': pq: unsupported comparison: bytes to crdb_internal_region
    --- FAIL: TestMrSystemDatabase/Sqlliveness_system_database (0.12s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMrSystemDatabase.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/ccl/multiregionccl/multiregion_system_table_test.go
+++ b/pkg/ccl/multiregionccl/multiregion_system_table_test.go
@@ -87,6 +87,7 @@ func TestMrSystemDatabase(t *testing.T) {
 
 	sDB := sqlutils.MakeSQLRunner(systemSQL)
 
+	sDB.Exec(t, `ANALYZE system.sqlliveness;`)
 	sDB.Exec(t, `SET CLUSTER SETTING sql.multiregion.system_database_multiregion.enabled = true`)
 	sDB.Exec(t, `ALTER DATABASE system SET PRIMARY REGION "us-east1"`)
 	sDB.Exec(t, `ALTER DATABASE system ADD REGION "us-east2"`)

--- a/pkg/sql/region_util.go
+++ b/pkg/sql/region_util.go
@@ -2521,6 +2521,10 @@ func (p *planner) setSystemDatabaseSurvival(ctx context.Context) error {
 	return nil
 }
 
+// systemTableLocalityChangeJobName is the job name used when the locality is converted
+// to regional by row.
+const systemTableLocalityChangeJobName = "convert system.%s to %s locality"
+
 // optimizeSystemDatabase configures some tables in the system data as
 // global and regional by row. The locality changes reduce how long it
 // takes a server to start up in a multi-region deployment.
@@ -2604,7 +2608,7 @@ func (p *planner) optimizeSystemDatabase(ctx context.Context) error {
 			return err
 		}
 
-		jobName := fmt.Sprintf("convert system.%s to %s locality", desc.GetName(), locality)
+		jobName := fmt.Sprintf(systemTableLocalityChangeJobName, desc.GetName(), locality)
 		return p.writeSchemaChange(ctx, desc, descpb.InvalidMutationID, jobName)
 	}
 

--- a/pkg/sql/schema_changer.go
+++ b/pkg/sql/schema_changer.go
@@ -770,7 +770,7 @@ func (sc *SchemaChanger) checkForMVCCCompliantAddIndexMutations(
 //
 // If the txn that queued the schema changer did not commit, this will be a
 // no-op, as we'll fail to find the job for our mutation in the jobs registry.
-func (sc *SchemaChanger) exec(ctx context.Context) error {
+func (sc *SchemaChanger) exec(ctx context.Context) (retErr error) {
 	sc.metrics.RunningSchemaChanges.Inc(1)
 	defer sc.metrics.RunningSchemaChanges.Dec(1)
 
@@ -920,7 +920,18 @@ func (sc *SchemaChanger) exec(ctx context.Context) error {
 	if sc.mutationID == descpb.InvalidMutationID {
 		// Nothing more to do.
 		isCreateTableAs := tableDesc.Adding() && tableDesc.IsAs()
-		return waitToUpdateLeases(isCreateTableAs /* refreshStats */)
+		// If we are converting a system database table to MR, we should force
+		// a stats refresh so that the stats also have the correct type. There is
+		// a potential race condition between waiting for leases and deleting the
+		// statistics using optimizeDatabase, where between those two point statistics
+		// with the wrong type could show up. To minimize any transient condition,
+		// lets force a refresh of stats here.
+		isSystemDatabaseTransformation := tableDesc.GetParentID() == keys.SystemDatabaseID &&
+			(strings.HasPrefix(sc.job.Payload().Description,
+				fmt.Sprintf(systemTableLocalityChangeJobName, tableDesc.GetName(), "regional by row")) ||
+				strings.HasPrefix(sc.job.Payload().Description,
+					fmt.Sprintf(systemTableLocalityChangeJobName, tableDesc.GetName(), "global")))
+		return waitToUpdateLeases(isCreateTableAs || isSystemDatabaseTransformation /* refreshStats */)
 	}
 
 	if err := sc.initJobRunningStatus(ctx); err != nil {
@@ -941,7 +952,7 @@ func (sc *SchemaChanger) exec(ctx context.Context) error {
 	}
 
 	defer func() {
-		if err := waitToUpdateLeases(err == nil /* refreshStats */); err != nil && !errors.Is(err, catalog.ErrDescriptorNotFound) {
+		if err := waitToUpdateLeases(retErr == nil /* refreshStats */); err != nil && !errors.Is(err, catalog.ErrDescriptorNotFound) {
 			// We only expect ErrDescriptorNotFound to be returned. This happens
 			// when the table descriptor was deleted. We can ignore this error.
 

